[15.0][FIX] hr_employee_document: Remove record rules from hr.employee to avoid side effects..

victoralmau opened the pull request:
Remove record rules from hr.employee to avoid side effects.

**Example of use case**:
- An employee X is created and a Manager Y (`parent_id`) is defined.
- User X creates an expense report (`hr.expense.sheet`).
- The "_Submit to manager_" button is clicked.
- A permissions error is displayed when accessing employee Y (`hr.employee`) because the record rules that this module had added do not allow access to that record.

**Solution**:
This change overrides some things to allow the user's employee attachments to be displayed.

Please @pedrobaeza can you review it?

@Tecnativa TT44536

pedrobaeza commented:
You have to put the steps to reproduce the problem in the commit message.

Apart from that, why that ACLs was there? (isn't they record rules, not ACL?) Seeing a bit the code, it seems you are solving it acting at code level. Please explain all of this in the commit message, docstring, etc.

victoralmau commented:
Sorry, changed the commit message and explained a use case.

pedrobaeza commented on hr_employee_document/models/ir_attachment.py:
```diff
@@ -0,0 +1,67 @@
+    _inherit = "ir.attachment"
+
+    @api.model
+    def _search(
```
Security shouldn't be handled in `_search` method. This is a security hole, as you can browse directly some random records.

victoralmau committed:
[FIX] hr_employee_document: Remove record rules from hr.employee to avoid side effects.

The purpose of this module is to allow a basic user (without hr permissions) to view
the attachments related to his employee (hr.employee).

By default a basic user cannot access to hr.employee model and therefore
cannot see the attachments (ir.attachment) linked to it.

This change overrides some things to allow the user's employee attachments to be displayed.

TT44536

victoralmau commented:
I have changed the approach to try to solve the problem in another way, although there is still some detail to be solved, what do you think about this way?

pedrobaeza commented on hr_employee_document/models/hr_employee.py:
```diff
@@ -22,15 +23,33 @@ def _compute_document_count(self):
+            and self._name == "hr.employee"
+        ):
+            if (
+                self.env.context.get("search_attachments_from_hr_employee")
```
Shouldn't this be limited to the specific employee?

pedrobaeza commented on hr_employee_document/models/ir_rule.py:
```diff
@@ -0,0 +1,38 @@
+        model to restrict only the user's employees when search employee attachments."""
+        res = super()._compute_domain(model_name, mode=mode)
+        user = self.env.user
+        if model_name == "hr.employee" and not self.env.su:
```
Shouldn't this also exclude HR managers?

victoralmau replied:
Now the extra domain is not added to the rule if the user has the Employees > Officer permission (this is the same behavior as the module currently has).

pedrobaeza replied:
But our use case wants to exclude Officers. We want this only for HR managers, not officers.

chienandalu reviewed (approved):
LGTM :+1:

pedrobaeza commented:
/ocabot merge major
